Handle non graceful node shutdown

iPraveenParihar opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

## Problem

When `ControllerUnpublishVolume` is called without the node first having
cleaned up the volume (via `NodeUnpublishVolume` and `NodeUnstageVolume`),
the CSI driver has no opportunity to revoke the node's access to the volume.
The node may still hold active mounts, open file handles, or client sessions.
This can lead to data corruption as applications may still be running on the
broken node with active client sessions, even though the node is marked as out
of service.

## Proposed solution

To ensure safe volume reuse and prevent stale client access during node
disruptions, the proposed solution is to track the client address during
the `NodeStageVolume()` operation and store it in the image or subvolume
metadata under the key: `.[rbd|cephfs].csi.ceph.com/clientAddress/<NodeId>`.

> 🔷 Note: This metadata should not be copied when creating clones, snapshots,
> or mirror images of the volume.
This stored address can then be used:

* In `ControllerUnpublishVolume()` to blocklist the client if the node
  has the `out-of-service` taints, ensuring it no longer has access to the
  volume.

* In `ControllerPublishVolume()` to remove the client from the blocklist,
  allowing the volume to be accessed again when the node recovers.

Ref design doc: https://github.com/ceph/ceph-csi/pull/5409

**Checklist:**

* [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [x] Reviewed the developer guide on [Submitting a Pull Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [ ] [Pending release notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next major release.
* [ ] Documentation has been updated, if necessary.
* [ ] Unit tests have been added, if necessary.
* [ ] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

iPraveenParihar committed:
api: add ControllerPublishSecretRef in csi-config-map

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
rbd: set clientAddress in image-metadata during NodeStageVolume

This commmit sets client address in image metadata. The client
address is fetched from connection.GetAddrs() which returns the
address of the current RADOS session which can be then used for
blocklisting.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
util: move ceph osd blocklist [add|rm] methods

This commit moves the ceph osd blocklist [add|rm] methods
to util/cephcmds.go for common usage.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
util: add k8s helper methods GetSecret and IsNodeOutOfService

This commit introduces:
- GetSecret to fetch k8s secret data
- IsNodeOutOfService to check for `out-of-service` taints on a node.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
rbd: implement Controller(Un)PublishVolume RPCs

This commit adds node fencing capabilities to RBD volumes in Kubernetes
environments:

1. ControllerUnpublishVolume(): Blocklists the client if the node has
   out-of-service taints, preventing access to the volume during node
   failures.

2. ControllerPublishVolume(): Removes the client from the blocklist,
   restoring volume access when the node recovers.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
cephfs: set clientAddress in subvolume-metadata during NodeStageVolume

This commmit sets client address in subvolume metadata. The client
address is fetched from connection.GetAddrs() which returns the
address of the current RADOS session which can be then used for
blocklisting.

Signed-off-by: Praveen M <m.praveen@ibm.com>

nixpanic commented on internal/util/k8s/node.go:
```diff
@@ -20,9 +20,14 @@ import (
 )
 
+const (
+	TaintOutOfServiceKey string = "node.kubernetes.io/out-of-service"
```
no need to export this outside of the package?

iPraveenParihar replied:
Done

nixpanic commented on internal/csi-addons/networkfence/fencing.go:
```diff
@@ -248,30 +248,7 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
-	log.DebugLog(ctx, "blocklisted IP %q successfully", ip)
-
-	return nil
+	return util.AddCephBlocklist(ctx, nf.Monitors, nf.cr, ip, useRange)
```
The message of this commit isn't really right, functions are moved for _some_ reason. Please mention that in the commit message.

iPraveenParihar replied:
Done

nixpanic commented on internal/util/cephcmds.go:
```diff
@@ -31,8 +31,11 @@ import (
-const InvalidPoolID int64 = -1
+const (
+	// InvalidPoolID used to denote an invalid pool.
+	InvalidPoolID int64 = -1
```
Does `InvalidPoolID` need to be public?

iPraveenParihar replied:
`InvalidPoolID` is referenced outside the file and its been public since then. This change is not part of this PR.

nixpanic replied:
Ah, right, missed that it was already there.

nixpanic commented on internal/util/cephcmds.go:
```diff
@@ -294,3 +297,62 @@ func RemoveObject(ctx context.Context, monitors string, cr *Credentials, poolNam
 	return nil
 }
+
+func AddCephBlocklist(ctx context.Context, monitors string, cr *Credentials, ip string, useRange bool) error {
```
this is now a shared function, please provide comments for it

iPraveenParihar replied:
Done

nixpanic commented on internal/util/k8s/node.go:
```diff
@@ -37,3 +42,26 @@ func GetNodeLabels(nodeName string) (map[string]string, error) {
 	return node.GetLabels(), nil
 }
+
+func IsNodeOutOfService(nodeName string) (bool, error) {
```
new functions should have comments

iPraveenParihar replied:
Done

nixpanic commented on internal/util/k8s/secrets.go:
```diff
@@ -0,0 +1,43 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func GetSecret(secretName, secretNamespace string) (map[string]string, error) {
```
Secrets are used elsewhere (KMS), maybe it provides a function like this too?

Adding helper functions like this, and updating other parts of the code to use them (remove duplication) can be done in their own PR. It'll make your main feature PR a little smaller and simpler to review.

iPraveenParihar replied:
Opened a tracker to address later once this PR is merged - https://github.com/ceph/ceph-csi/issues/5431

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -397,6 +403,50 @@ func (ns *NodeServer) NodeStageVolume(
+
+	// The address we get is 10.244.0.1:0/2686266785 from
+	// which we need to extract the IP address.
+	ipAddress, err := nf.ParseClientIP(address)
```
If `csi-addons/networkfence` is imported because of `ParseClientIP`, move the function to `internal/util/.../something`.

iPraveenParihar replied:
Done

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -87,6 +87,9 @@ const (
 	clusterNameKey = "csi.ceph.com/cluster/name"
 
+	// clientAddressKey is the key used to store the client address.
+	clientAddressKey = ".rbd.csi.ceph.com/clientAddress"
```
why start with `.`? Might be good to explain that in the comment.

iPraveenParihar replied:
starts with `.` to avoid copying it to the mirrored image.
Done

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -282,6 +285,12 @@ var (
 
+// GetClientAddressKey returns the key to store the client address in the
+// image metadata.
+func GetClientAddressKey(nodeId string) string {
```
this feels like a non-RBD specific util function, it does not belong in the `rbd` package, or should not be a public function

iPraveenParihar replied:
`GetClientAddressKey` constructs the metadata key for the clientAddress.
takes in `nodeId` and appends to RBD image specific key (`.rbd.csi.ceph.com/clientAddress`)

I'll make it a private function.

nixpanic commented on internal/cephfs/controllerserver.go:
```diff
@@ -1119,3 +1119,222 @@ func deleteSnapshotAndUndoReservation(
+		return nil, status.Error(codes.InvalidArgument, "Node ID cannot be empty")
+	}
+	secrets := req.GetSecrets()
+	if k8s.RunsOnKubernetes() {
```
make this
```go
if !k8s.RunsOnKubernetes() {
    return ...
}
```

keeps the function a little cleaner (less indention)

iPraveenParihar replied:
Done,

nixpanic commented on internal/cephfs/controllerserver.go:
```diff
@@ -1119,3 +1119,222 @@ func deleteSnapshotAndUndoReservation(
+	return secrets, nil
+}
+
+func (cs *ControllerServer) fenceNode(
```
please describe briefly what steps this function takes to fence a node

iPraveenParihar replied:
done

nixpanic commented on internal/cephfs/nodeserver.go:
```diff
@@ -326,6 +334,38 @@ func (ns *NodeServer) NodeStageVolume(
 	return &csi.NodeStageVolumeResponse{}, nil
 }
 
+func setClientAddress(ctx context.Context, nodeId string, volOptions *store.VolumeOptions) error {
```
comment needed that mentions were/how the client address is set. Maybe rename `nodeId` to `clientAddress`?

iPraveenParihar replied:
nodeId is the nodeName, clientAddress is fetched in `setClientaddress()` function.
added comment.

nixpanic commented on internal/cephfs/nodeserver.go:
```diff
@@ -326,6 +334,38 @@ func (ns *NodeServer) NodeStageVolume(
 	return &csi.NodeStageVolumeResponse{}, nil
 }
 
+func setClientAddress(ctx context.Context, nodeId string, volOptions *store.VolumeOptions) error {
```
If this is the _set_ function, does there need to be an _unset_ function too?

iPraveenParihar replied:
`setClientAddress` includes fetching the clientAddress and store in metadata and unset is directly done using the `RemoveMetadata(<metadata-key>)` function in `Controller(Un)PublishVolume()` call

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -1694,20 +1693,197 @@ func (cs *ControllerServer) ControllerPublishVolume(
-		PublishContext: map[string]string{},
-	}, nil
+	secrets := req.GetSecrets()
+	if k8s.RunsOnKubernetes() {
```
here as well
```go
if !k8s.RunsOnKubernetes() {
```

iPraveenParihar replied:
done

nixpanic reviewed (commented):
Might be easier to first do one of CephFS/RBD and then the other?

iPraveenParihar committed:
util: move ParseClientIP to internal/util/util.go

This commit moves ParseClientIP to internal/util/util.go
for common usage.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar commented:
Thanks for the reviews, I have addressed them.

> Might be easier to first do one of CephFS/RBD and then the other?

Since, the logic is similar for both CephFS and RBD thought of having in same PR.

Rakshith-R reviewed (commented):
LGTM,
Just few changes.

Please add 
- a flag to guard this behavior as well defaulting to false to prevent breaking upgraded clusters.
- release note

The following can be added in a follow-up pr too
- documentation in 
  - https://github.com/ceph/ceph-csi/blob/devel/deploy/csi-config-map-sample.yaml 
  - usage steps doc:
    - https://github.com/ceph/ceph-csi/blob/devel/docs/rbd/deploy.md
    - https://github.com/ceph/ceph-csi/blob/devel/docs/cephfs/deploy.md

iPraveenParihar committed:
util: add enable-fencing flag

Introduce a new flag 'enable-fencing' false by default.
When enabled driver can set the client address per node in
the volume metadata during NodeStageVolume operation and
fence the node in ControllerUnpublishVolume and unfence
the node in ControllerPublishVolume using the address stored
in the volume metadata.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar commented on internal/cephfs/nodeserver.go:
```diff
@@ -326,6 +334,49 @@ func (ns *NodeServer) NodeStageVolume(
+		return fmt.Errorf("failed to parse client address: %w", err)
+	}
+
+	subvolumeClient := core.NewSubVolume(conn, &volOptions.SubVolume, volOptions.ClusterID, "", true)
```
@Madhu-1, here I have passed true for `setMetadata`.  I'm quite confused with `--setmetadata` flag. 
Is it only for CephFS or RBD as well?

I see we set `rv.EnableMetadata` in ControllerServer, but at most places we never use it unlike CephFS where we have checks before set/unset metadata.

If the whole purpose of the flag was to avoid failures on Ceph version which didn't support metadata operations then can we remove/deprecated the flag now?

Madhu-1 replied:
This is already enabled by default in last release. Yes we can deprecate the flag in next release

iPraveenParihar replied:
Okay, tracker - https://github.com/ceph/ceph-csi/issues/5433

Madhu-1 commented on cmd/cephcsi.go:
```diff
@@ -75,6 +75,7 @@ func init() {
 	flag.StringVar(&conf.StagingPath, "stagingpath", defaultStagingPath, "staging path")
 	flag.StringVar(&conf.ClusterName, "clustername", "", "name of the cluster")
 	flag.BoolVar(&conf.SetMetadata, "setmetadata", true, "set metadata on the volume")
+	flag.BoolVar(&conf.SetAddressMetadata, "setaddressmetadata", false, "set client address metadata on the volume")
```
We should use different key and description should be about BlockListing/Unblocklisting.

iPraveenParihar replied:
How about
```
flag.BoolVar(&conf.EnableFencing, "enable-fencing", false, "enable fencing of nodes during non-graceful shutdowns.")
```
?

Madhu-1 replied:
@iPraveenParihar helm and template update required for this new field. can you open a tracker to add the support for it?

iPraveenParihar commented:
@Rakshith-R, I have added the flag `--setaddressmetadata` false by default. Which performs the fencing/un-fencing operation only when enabled. PTAL

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -1119,3 +1119,286 @@ func deleteSnapshotAndUndoReservation(
+	// This is because the unfencing logic is specific to Kubernetes environments.
+	if !k8s.RunsOnKubernetes() {
+		return &csi.ControllerPublishVolumeResponse{}, nil
+	}
```
Move thsi to top and we can avoid validating the Input and taking locks unnecessary

iPraveenParihar replied:
done

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -1119,3 +1119,286 @@ func deleteSnapshotAndUndoReservation(
+	// If the CSI driver is not running on Kubernetes, we do not need to
+	// fence the node, so we return immediately.
+	// This is because the fencing logic is specific to Kubernetes environments.
+	if !k8s.RunsOnKubernetes() {
```
same here as well.

iPraveenParihar replied:
done

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -1119,3 +1119,286 @@ func deleteSnapshotAndUndoReservation(
+
+// getControllerPublishSecret retrieves the controller publish secret
+// from the CSI config file for the given cluster ID.
+func getControllerPublishSecret(clusterID string) (map[string]string, error) {
```
Add unit test for this function?

Madhu-1 replied:
Can we move this to until and have common function for both cephfs and rbd to avoid dplications?

iPraveenParihar replied:
> Can we move this to until and have common function for both cephfs and rbd to avoid dplications?

`getControllerPublishSecret` internally calls RBD/CephFS specific calls. To know, will need to pass some decision making parameter. 

lets keep it simple for now?

iPraveenParihar replied:
[Update]
we already have [rbd|cephfs|nfs]Type in cephcsi.go
now moved to util and common function for `getControllerPublishSecret`

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -1119,3 +1119,286 @@ func deleteSnapshotAndUndoReservation(
+	return secrets, nil
+}
+
+// fenceNode attempts to fence a client node from accessing the RBD volume.
```
cephfs

iPraveenParihar replied:
done

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -1119,3 +1119,286 @@ func deleteSnapshotAndUndoReservation(
+
+		secrets, err = getControllerPublishSecret(vi.ClusterID)
+		if err != nil {
+			log.WarningLog(ctx, "controller publish secret not found: %w", err)
```
%s

iPraveenParihar replied:
done

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -1119,3 +1119,286 @@ func deleteSnapshotAndUndoReservation(
+
+		secrets, err = getControllerPublishSecret(vi.ClusterID)
+		if err != nil {
+			log.WarningLog(ctx, "controller publish secret not found: %w", err)
```
%s

iPraveenParihar replied:
done

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -1119,3 +1119,286 @@ func deleteSnapshotAndUndoReservation(
+
+	clientAddress := metadataValues[metadataKey]
+	if clientAddress == "" {
+		log.WarningLog(ctx, "client address metadata %s for volume %s is empty", metadataKey, volOptions.VolID)
```
This will log false details even if the key is not present in the map

iPraveenParihar replied:
Changed to
- If doesn't exists, return nil
- If exists and empty, return error

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -1119,3 +1119,286 @@ func deleteSnapshotAndUndoReservation(
+	secrets map[string]string,
+	setAddressMetadata bool,
+) error {
+	if !setAddressMetadata {
```
There is duplicate code in the initialization part of both fence and unfence, we can move it to helper to avoid duplication?

iPraveenParihar replied:
I tried it, but kinda becomes more complex and ugly.
Can we take this in enhancement PR?

Madhu-1 replied:
@iPraveenParihar can you please open a tracker for it.

Madhu-1 commented on internal/cephfs/core/metadata.go:
```diff
@@ -27,11 +27,21 @@ import (
 	clusterNameKey = "csi.ceph.com/cluster/name"
+
+	// clientAddressKey is the key used to store the client address.
+	// starts with `.` to avoid copying it to the mirrored image.
```
subvolume?

iPraveenParihar replied:
done

Madhu-1 commented on internal/cephfs/core/metadata.go:
```diff
@@ -149,3 +175,25 @@ func (s *subVolumeClient) UnsetAllMetadata(keys []string) error {
+
+	metadataValues := make(map[string]string)
+	for _, key := range keys {
+		value, err := s.getMetadata(key)
```
This is not nice to have, It would be good to use List instead of making multiple ceph calls to get a key each time.

iPraveenParihar replied:
Implemented ListMetadata

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -282,6 +285,11 @@ func (ns *NodeServer) NodeStageVolume(
+	err = setClientAddress(ctx, ns.Driver.GetNodeID(), volOptions, ns.Driver.IsSetAddressMetadataEnabled())
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
```
If the volume is already moved we are returning the success above. we might miss setting the metadata in corner cases

iPraveenParihar replied:
missed it, now moved above

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -326,6 +334,49 @@ func (ns *NodeServer) NodeStageVolume(
+	}
+	if nodeId == "" {
+		return errors.New("nodeId is empty")
+	}
```
When this can happen? i think we already have validation for nodeID, if not this should be done at the driver level not here.

iPraveenParihar replied:
I missed it, we are already have validation for empty NodeId

https://github.com/ceph/ceph-csi/blob/1f2706fd345bb42e20f520f17aa924dfd872f1fa/internal/csi-common/driver.go#L48-L59

removed from here

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -326,6 +334,49 @@ func (ns *NodeServer) NodeStageVolume(
+		return fmt.Errorf("failed to get client address: %w", err)
+	}
+
+	log.DebugLog(ctx, "cephfs: setting client address %s for subvolume %s", address, volOptions.VolID)
```
for volumeID instead of subvolume as we are logging the volID not the subvolume  name?

Madhu-1 replied:
Lets avoid this log and log at the end of this function that includes the address as well.

iPraveenParihar replied:
> for volumeID instead of subvolume as we are logging the volID not the subvolume name?

Changed to `volOptions.SubVolume.VolID`

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1694,20 +1694,265 @@ func (cs *ControllerServer) ControllerPublishVolume(
+	// If the CSI driver is not running on Kubernetes, we do not need to
+	// unfence the node, so we return immediately.
+	// This is because the unfencing logic is specific to Kubernetes environments.
+	if !k8s.RunsOnKubernetes() {
+		return &csi.ControllerPublishVolumeResponse{}, nil
+	}
```
Return early to avoid locking?

iPraveenParihar replied:
done

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1694,20 +1694,265 @@ func (cs *ControllerServer) ControllerPublishVolume(
 	if req.GetVolumeId() == "" {
 		return nil, status.Error(codes.InvalidArgument, "Volume ID cannot be empty")
 	}
+	if req.GetNodeId() == "" {
+		return nil, status.Error(codes.InvalidArgument, "Node ID cannot be empty")
+	}
```
NodeId is optional the RPC implementation, we should remove it for all nodeID's (this is for future)

iPraveenParihar replied:
> we should remove it for all nodeID's (this is for future)

you mean if NodeId is empty, we remove fencing for all NodeId?

Madhu-1 replied:
I have added a comment for cephfs for this one. if NodeID is empty lets return success and dont fencing anything as we dont know what node need to be fenced

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1694,20 +1694,265 @@ func (cs *ControllerServer) ControllerPublishVolume(
+	if clientAddress == "" {
+		// If the metadata is empty, we do not need to add it to the blocklist
+		// as it does not contain any client address to block.
+		log.WarningLog(ctx, "metadata %s for image %s is empty", metadataKey, rv)
```
Lets avoid unwanted warning and have a debug log of we are moving to add blocklisting?

iPraveenParihar replied:
done

Madhu-1 commented on internal/util/csiconfig.go:
```diff
@@ -241,3 +241,29 @@ func GetCephFSMountOptions(pathToConfig, clusterID string) (string, string, erro
+
+// GetRBDControllerPublishSecretRef returns the secret name and namespace used for
+// controller publish operations for RBD volumes.
+func GetRBDControllerPublishSecretRef(pathToConfig, clusterID string) (string, string, error) {
+	cluster, err := readClusterInfo(pathToConfig, clusterID)
+	if err != nil {
+		return "", "", err
+	}
+
+	secretRef := cluster.RBD.ControllerPublishSecretRef
+
+	return secretRef.Name, secretRef.Namespace, nil
+}
+
+// GetCephFSControllerPublishSecretRef returns the secret name and namespace used for
+// controller publish operations for CephFS volumes.
+func GetCephFSControllerPublishSecretRef(pathToConfig, clusterID string) (string, string, error) {
```
Add unit test for this one.

iPraveenParihar replied:
done

Madhu-1 commented on internal/util/k8s/secrets.go:
```diff
@@ -0,0 +1,46 @@
+		return nil, fmt.Errorf("failed to connect to Kubernetes: %w", err)
+	}
+
+	secret, err := client.CoreV1().Secrets(secretNamespace).Get(context.TODO(), secretName, metav1.GetOptions{})
```
pass the context instead of creating one here?

iPraveenParihar replied:
all k8s helper functions have context.TODO(), I'll cleanup that https://github.com/ceph/ceph-csi/issues/5431

nixpanic replied:
please note this concern in that issue too

Madhu-1 commented on internal/util/k8s/node.go:
```diff
@@ -37,3 +42,29 @@ func GetNodeLabels(nodeName string) (map[string]string, error) {
+			"to connect to Kubernetes: %w", nodeName, err)
+	}
+
+	node, err := client.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
```
pass the context instead of creating one?

nixpanic replied:
still TODO?

Madhu-1 reviewed (changes requested):
overall its in good state.

iPraveenParihar committed:
cephfs: implement Controller(Un)publishVolume RPCs

This commit adds node fencing capabilities to CephFS
subvolumes in Kubernetes environments:

1. ControllerUnpublishVolume(): Blocklists the client if the node has
   out-of-service taints, preventing access to the volume during node
   failures.

2. ControllerPublishVolume(): Removes the client from the blocklist,
   restoring volume access when the node recovers.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
cleanup: move driverType to internal/util

Relocates driver type constants to internal/util for shared usage
across the codebase, establishing a single source of truth for
driver type identification.

Signed-off-by: Praveen M <m.praveen@ibm.com>

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -1119,3 +1119,267 @@ func deleteSnapshotAndUndoReservation(
+	}
+	defer cs.VolumeLocks.Release(req.GetVolumeId())
+
+	err := cs.unfenceNode(ctx, req.GetNodeId(), req.GetVolumeId(), secrets, cs.Driver.IsFencingEnabled())
```
instead of `secrets` directly call req.GetSecrets() here

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -1119,3 +1119,267 @@ func deleteSnapshotAndUndoReservation(
 	return nil
 }
+
+// ControllerPublishVolume implements the CSI ControllerPublishVolume RPC.
```
This is generic comment, can it be updated with little details

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -1119,3 +1119,267 @@ func deleteSnapshotAndUndoReservation(
+	return &csi.ControllerPublishVolumeResponse{}, nil
+}
+
+// ControllerUnpublishVolume implements the CSI ControllerUnpublishVolume RPC.
```
same here as well

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -1119,3 +1119,267 @@ func deleteSnapshotAndUndoReservation(
+		return nil, status.Error(codes.InvalidArgument, "Volume ID cannot be empty")
+	}
+	// NodeId is optional in the ControllerUnpublishVolumeRequest.
+	if req.GetNodeId() == "" {
+		return nil, status.Error(codes.InvalidArgument, "Node ID cannot be empty")
+	}
```
Am not sure this is correct check as it can have side effects in case CO doesnt set it as this field is optional. How about we dont do fencing if this is empty and return success from here as we cannot do fencing?( This is to keep today behavior and to be on safer side)

iPraveenParihar replied:
Lets, remove the check from here and add a check in `fenceNode()`

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -1119,3 +1119,267 @@ func deleteSnapshotAndUndoReservation(
+	}
+	defer cs.VolumeLocks.Release(req.GetVolumeId())
+
+	err := cs.fenceNode(ctx, req.GetNodeId(), req.GetVolumeId(), secrets, cs.Driver.IsFencingEnabled())
```
call req.GetSecrets()

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -1119,3 +1119,267 @@ func deleteSnapshotAndUndoReservation(
+	// If the CSI driver is not running on Kubernetes, we do not need to
+	// fence the node, so we return immediately.
+	// This is because the fencing logic is specific to Kubernetes environments.
+	if !k8s.RunsOnKubernetes() {
```
can we have check for `cs.Driver.IsFencingEnabled()` here only so that we dont need to processed for validation as the feature is disabled?

iPraveenParihar replied:
Today, we just have fencing operations but we might have other things in future?
For example, removing `nodeId:userId` mapping metadata in ControllerUnpublishVolume.

Madhu-1 replied:
sounds good, i missed that point

Madhu-1 commented on internal/cephfs/core/metadata.go:
```diff
@@ -149,3 +177,22 @@ func (s *subVolumeClient) UnsetAllMetadata(keys []string) error {
+	if errors.Is(err, ErrSubVolMetadataNotSupported) {
+		return nil, nil
+	}
+	if err != nil && !errors.Is(err, libcephfs.ErrNotExist) {
+		return nil, fmt.Errorf("failed to list metadata on subvolume %v: %w", s, err)
```
why do we need extra check for `libcephfs.ErrNotExist` this should return error in anycase not nil when subvolume is not found

Madhu-1 commented on internal/cephfs/driver.go:
```diff
@@ -127,14 +127,16 @@ func (fs *Driver) Run(conf *util.Config) {
 	if conf.IsControllerServer || !conf.IsNodeServer {
 		fs.cd.AddControllerServiceCapabilities([]csi.ControllerServiceCapability_RPC_Type{
 			csi.ControllerServiceCapability_RPC_CREATE_DELETE_VOLUME,
+			csi.ControllerServiceCapability_RPC_PUBLISH_UNPUBLISH_VOLUME,
```
question: If fencing is disabled do we need to advertise the `ControllerServiceCapability_RPC_PUBLISH_UNPUBLISH_VOLUME` cap?

iPraveenParihar replied:
we need it for nodeID:userId mapping metadata as well.

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -326,6 +333,46 @@ func (ns *NodeServer) NodeStageVolume(
+// The connection.GetAddrs() method returns an address in the format "10.244.0.1:0/2686266785".
+// We parse this to extract just the IP address portion (e.g., "10.244.0.1") and store in metadata.
+// If '--enable-fencing' flag is set to false in CSI driver configuration, this function does nothing.
+func setClientAddress(
```
Make this as method for ns, so that we dont need to pass the NodeID and IsFencingEnabled

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1694,20 +1701,236 @@ func (cs *ControllerServer) ControllerPublishVolume(
+		// If the node is not tainted with `out-of-service`,
+		// we remove the client address metadata from the volume metadata.
+		err = rv.RemoveMetadata(metadataKey)
+		if err != nil && !errors.Is(err, librbd.ErrNotFound) {
```
Do we get ErrNotExist or ErrNotFound error here?

iPraveenParihar replied:
`ErrNotExist` thanks!

Madhu-1 commented on internal/rbd/driver/driver.go:
```diff
@@ -102,13 +102,15 @@ func (r *Driver) Run(conf *util.Config) {
 	if conf.IsControllerServer || !conf.IsNodeServer {
 		r.cd.AddControllerServiceCapabilities([]csi.ControllerServiceCapability_RPC_Type{
 			csi.ControllerServiceCapability_RPC_CREATE_DELETE_VOLUME,
+			csi.ControllerServiceCapability_RPC_PUBLISH_UNPUBLISH_VOLUME,
```
dont advertise this Cap is the fencing is disabled?

Madhu-1 commented on internal/rbd/nodeserver.go:
```diff
@@ -397,6 +402,59 @@ func (ns *NodeServer) NodeStageVolume(
+// The connection.GetAddrs() method returns an address in the format "10.244.0.1:0/2686266785".
+// We parse this to extract just the IP address portion (e.g., "10.244.0.1") and store in metadata.
+// If '--enable-fencing' flag is set to false in CSI driver configuration, this function does nothing.
+func (rv *rbdVolume) setClientAddress(
```
this should not be a method on the rbd it should be on `ns`?

Madhu-1 commented on internal/rbd/nodeserver.go:
```diff
@@ -397,6 +402,59 @@ func (ns *NodeServer) NodeStageVolume(
+	if !enableFencing {
+		return nil
+	}
+	if nodeId == "" {
+		return errors.New("node ID is empty")
+	}
```
we dont need this check as the NodeID is already validated in `NewCSIDriver` function

iPraveenParihar replied:
Yea, missed removing it

Madhu-1 reviewed (changes requested):
There are few missing parts that is required for this feature.

* cephfs deployment doesnt include the attacher
* cephfs CSI driver object attacher is set to false
* Add this feature to Pending Release notes and document that all cephfs deployment should include the csi-attacher

We are almost there, Thanks for address all the comments. great work 🎉

iPraveenParihar committed:
doc: update pendingReleaseNotes

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar commented:
> There are few missing parts that is required for this feature.
> 
> * cephfs deployment doesnt include the attacher

do we need to update cephfs templates in ceph-csi?

> * cephfs CSI driver object attacher is set to false

you mean in the ceph-csi csidriver.yaml files?

> * Add this feature to Pending Release notes and document that all cephfs deployment should include the csi-attacher
> 
> We are almost there, Thanks for address all the comments. great work 🎉


Thanks for the review, appreciated 😄

Madhu-1 commented on deploy/cephfs/kubernetes/csidriver.yaml:
```diff
@@ -11,7 +11,7 @@ kind: CSIDriver
   name: "cephfs.csi.ceph.com"
 spec:
-  attachRequired: false
+  attachRequired: true
```
Do we have attacher running in the cephfs deployment?

iPraveenParihar replied:
In ceph-csi standalone deployment (i.e, via deploy/ templates or helm) it was removed from CephFS deployment here https://github.com/ceph/ceph-csi/pull/3149.

But in Rook & Ceph-CSI-Operator attacher is deployed by default for CephFS deployment.

Madhu-1 replied:
we need to add it back or else we cannot use the functionality with cephfs with standalone cephcsi? Because we have not moved to the ceph-csi-operator yet!

iPraveenParihar replied:
Added attacher sidecar in CephFS deployments in helm templates and deploy/ templates.

Madhu-1 reviewed (commented):
LGTM. Thanks for all the work

Can you please run an E2E to see if everything is working as expected?

iPraveenParihar commented:
Testing covered

1. Feature disabled by default
```
I0721 05:29:35.130913       1 controllerserver-default.go:43] ID: 8 Using default ControllerGetCapabilities
I0721 05:29:35.131244       1 utils.go:348] ID: 8 GRPC response: {"capabilities":[{"rpc":{"type":"CREATE_DELETE_VOLUME"}},{"rpc":{"type":"CREATE_DELETE_SNAPSHOT"}},{"rpc":{"type":"CLONE_VOLUME"}},{"rpc":{"type":"EXPAND_VOLUME"}}]}
```

metadata is not stored - 
```
bash-5.1$ rbd image-meta ls replicapool/csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94
There are 3 metadata on this image:

Key                               Value
csi.storage.k8s.io/pv/name        pvc-f3173c10-61d4-4e43-b60f-a7b961e689fb
csi.storage.k8s.io/pvc/name       rbd-pvc
csi.storage.k8s.io/pvc/namespace  default
```
<hr>
2. Feature enabled (`--enable-fencing=true`)

/ControllerPublishVolume
```
I0721 05:38:41.724741       1 utils.go:341] ID: 24 GRPC call: /csi.v1.Controller/ControllerPublishVolume
I0721 05:38:41.731915       1 utils.go:342] ID: 24 GRPC request: {"node_id":"c1-m03","secrets":"***stripped***","volume_capability":{"access_mode":{"mode":"SINGLE_NODE_WRITER"},"block":{}},"volume_context":{"clusterID":"rook-ceph","imageFeatures":"layering","imageFormat":"2","imageName":"csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94","journalPool":"replicapool","pool":"replicapool","storage.kubernetes.io/csiProvisionerIdentity":"1753075775496-228-rook-ceph.rbd.csi.ceph.com"},"volume_id":"0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94"}
I0721 05:38:41.960430       1 omap.go:89] ID: 24 got omap values: (pool="replicapool", namespace="", name="csi.volume.1d186650-ea71-49b8-b98e-e68a22cb9f94"): map[csi.imageid:580777684f59f7 csi.imagename:csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94 csi.volname:pvc-f3173c10-61d4-4e43-b60f-a7b961e689fb csi.volume.owner:default]
W0721 05:38:42.174497       1 controllerserver.go:1904] ID: 24 metadata .rbd.csi.ceph.com/clientAddress/c1-m03 is not set for image replicapool/csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94
I0721 05:38:42.174822       1 utils.go:348] ID: 24 GRPC response: {}
```

/NodeStageVolume - metadata is set
```
I0721 05:38:51.498509 2957458 utils.go:341] ID: 10 Req-ID: 0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94 GRPC call: /csi.v1.Node/NodeStageVolume
I0721 05:38:51.499756 2957458 utils.go:342] ID: 10 Req-ID: 0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94 GRPC request: {"secrets":"***stripped***","staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/staging/pvc-f3173c10-61d4-4e43-b60f-a7b961e689fb","volume_capability":{"access_mode":{"mode":"SINGLE_NODE_MULTI_WRITER"},"block":{}},"volume_context":{"clusterID":"rook-ceph","imageFeatures":"layering","imageFormat":"2","imageName":"csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94","journalPool":"replicapool","pool":"replicapool","storage.kubernetes.io/csiProvisionerIdentity":"1753075775496-228-rook-ceph.rbd.csi.ceph.com"},"volume_id":"0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94"}
I0721 05:38:51.606647 2957458 omap.go:89] ID: 10 Req-ID: 0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94 got omap values: (pool="replicapool", namespace="", name="csi.volume.1d186650-ea71-49b8-b98e-e68a22cb9f94"): map[csi.imageid:580777684f59f7 csi.imagename:csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94 csi.volname:pvc-f3173c10-61d4-4e43-b60f-a7b961e689fb csi.volume.owner:default]
I0721 05:38:51.754304 2957458 rbd_util.go:382] ID: 10 Req-ID: 0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94 checking for ImageFeatures: [layering]
I0721 05:38:51.821824 2957458 nodeserver.go:450] ID: 10 Req-ID: 0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94 metadata .rbd.csi.ceph.com/clientAddress/c1-m03 set for image replicapool/csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94
I0721 05:38:51.963993 2957458 cephcmds.go:109] ID: 10 Req-ID: 0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94 command succeeded: rbd [device list --format=json --device-type krbd]
I0721 05:38:52.020378 2957458 rbd_attach.go:439] ID: 10 Req-ID: 0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94 rbd: map mon 10.105.101.174:6789
I0721 05:38:52.302822 2957458 cephcmds.go:109] ID: 10 Req-ID: 0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94 command succeeded: rbd [--id csi-rbd-node -m 10.105.101.174:6789 --keyfile=***stripped*** map replicapool/csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94 --device-type krbd --options noudev]
I0721 05:38:52.303061 2957458 nodeserver.go:479] ID: 10 Req-ID: 0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94 rbd image: replicapool/csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94 was successfully mapped at /dev/rbd0
I0721 05:38:52.303445 2957458 mount_linux.go:670] Attempting to determine if disk "/dev/rbd0" is formatted using blkid with args: ([-p -s TYPE -s PTTYPE -o export /dev/rbd0])
I0721 05:38:52.347595 2957458 mount_linux.go:673] Output: ""
I0721 05:38:52.348194 2957458 mount_linux.go:260] Mounting cmd (mount) with arguments ( -o bind,_netdev /dev/rbd0 /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/staging/pvc-f3173c10-61d4-4e43-b60f-a7b961e689fb/0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94)
I0721 05:38:52.359045 2957458 mount_linux.go:260] Mounting cmd (mount) with arguments ( -o bind,remount,_netdev /dev/rbd0 /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/staging/pvc-f3173c10-61d4-4e43-b60f-a7b961e689fb/0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94)
I0721 05:38:52.365783 2957458 nodeserver.go:396] ID: 10 Req-ID: 0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94 rbd: successfully mounted volume 0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94 to stagingTargetPath /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/staging/pvc-f3173c10-61d4-4e43-b60f-a7b961e689fb/0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94
I0721 05:38:52.366082 2957458 utils.go:348] ID: 10 Req-ID: 0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94 GRPC response: {}
```

```
bash-5.1$ rbd image-meta ls replicapool/csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94
There are 4 metadata on this image:

Key                                     Value
.rbd.csi.ceph.com/clientAddress/c1-m03  192.168.39.16
csi.storage.k8s.io/pv/name              pvc-f3173c10-61d4-4e43-b60f-a7b961e689fb
csi.storage.k8s.io/pvc/name             rbd-pvc
csi.storage.k8s.io/pvc/namespace        default
```

/ControllerUnpublishVolume - metadata is removed

```
I0721 05:43:58.791783       1 utils.go:341] ID: 25 GRPC call: /csi.v1.Controller/ControllerUnpublishVolume
I0721 05:43:58.793389       1 utils.go:342] ID: 25 GRPC request: {"node_id":"c1-m03","secrets":"***stripped***","volume_id":"0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94"}
I0721 05:43:58.805395       1 omap.go:89] ID: 25 got omap values: (pool="replicapool", namespace="", name="csi.volume.1d186650-ea71-49b8-b98e-e68a22cb9f94"): map[csi.imageid:580777684f59f7 csi.imagename:csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94 csi.volname:pvc-f3173c10-61d4-4e43-b60f-a7b961e689fb csi.volume.owner:default]
I0721 05:43:58.936552       1 controllerserver.go:1816] ID: 25 node is not marked as out-of-service, removed metadata .rbd.csi.ceph.com/clientAddress/c1-m03 from image replicapool/csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94 if it existed
I0721 05:43:58.936779       1 utils.go:348] ID: 25 GRPC response: {}
```

```
bash-5.1$ rbd image-meta ls replicapool/csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94
There are 3 metadata on this image:

Key                               Value
csi.storage.k8s.io/pv/name        pvc-f3173c10-61d4-4e43-b60f-a7b961e689fb
csi.storage.k8s.io/pvc/name       rbd-pvc
csi.storage.k8s.io/pvc/namespace  default
```
<hr>

3. Node marked for out-of-service

```
[pm@dhcp53-115 rbd]$ k get node
NAME     STATUS   ROLES           AGE   VERSION
c1       Ready    control-plane   20d   v1.32.0
c1-m02   Ready    worker          20d   v1.32.0
c1-m03   Ready    worker          12d   v1.32.0
[pm@dhcp53-115 rbd]$ k get po --field-selector spec.nodeName=c1-m03
NAME                       READY   STATUS    RESTARTS   AGE
rbd-demo-5d6f7d499-vj465   1/1     Running   0          81s
```

```
bash-5.1$ rbd status replicapool/csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94
Watchers:
	watcher=192.168.39.16:0/3530573161 client.5771273 cookie=18446462598732840961
bash-5.1$ rbd image-meta ls replicapool/csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94
There are 4 metadata on this image:

Key                                     Value
.rbd.csi.ceph.com/clientAddress/c1-m03  192.168.39.16
csi.storage.k8s.io/pv/name              pvc-f3173c10-61d4-4e43-b60f-a7b961e689fb
csi.storage.k8s.io/pvc/name             rbd-pvc
csi.storage.k8s.io/pvc/namespace        default
```

Node is unresponsive and Status: NotReady
```
[pm@dhcp53-115 rbd]$ k get node
NAME     STATUS     ROLES           AGE   VERSION
c1       Ready      control-plane   20d   v1.32.0
c1-m02   Ready      worker          20d   v1.32.0
c1-m03   NotReady   worker          12d   v1.32.0
```

```
[pm@dhcp53-115 rbd]$ k taint node c1-m03 node.kubernetes.io/out-of-service=nodeshutdown:NoExecute
node/c1-m03 tainted
```

node c1-m03 is blocklisted

/ControllerUnpublishVolume
```
I0721 05:57:44.945602       1 utils.go:341] ID: 32 GRPC call: /csi.v1.Controller/ControllerUnpublishVolume
I0721 05:57:44.945693       1 utils.go:342] ID: 32 GRPC request: {"node_id":"c1-m03","secrets":"***stripped***","volume_id":"0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94"}
I0721 05:57:44.954255       1 omap.go:89] ID: 32 got omap values: (pool="replicapool", namespace="", name="csi.volume.1d186650-ea71-49b8-b98e-e68a22cb9f94"): map[csi.imageid:580777684f59f7 csi.imagename:csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94 csi.volname:pvc-f3173c10-61d4-4e43-b60f-a7b961e689fb csi.volume.owner:default]
I0721 05:57:46.204542       1 cephcmds.go:109] ID: 32 command succeeded: ceph [osd blocklist add 192.168.39.16 157784760 --id=csi-rbd-provisioner --keyfile=***stripped*** -m=10.105.101.174:6789]
I0721 05:57:46.204616       1 cephcmds.go:325] ID: 32 blocklisted IP "192.168.39.16" successfully
I0721 05:57:46.204695       1 utils.go:348] ID: 32 GRPC response: {}
```

```
bash-5.1$ ceph osd blocklist ls
192.168.39.16:0/0 2030-07-21T11:03:45.748836+0000
listed 1 entries
```

Pod is scheduled to other node
/ControllerPublishVolume
```
I0721 05:57:46.589431       1 utils.go:341] ID: 33 GRPC call: /csi.v1.Controller/ControllerPublishVolume
I0721 05:57:46.589562       1 utils.go:342] ID: 33 GRPC request: {"node_id":"c1-m02","secrets":"***stripped***","volume_capability":{"access_mode":{"mode":"SINGLE_NODE_WRITER"},"block":{}},"volume_context":{"clusterID":"rook-ceph","imageFeatures":"layering","imageFormat":"2","imageName":"csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94","journalPool":"replicapool","pool":"replicapool","storage.kubernetes.io/csiProvisionerIdentity":"1753075775496-228-rook-ceph.rbd.csi.ceph.com"},"volume_id":"0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94"}
I0721 05:57:46.593552       1 omap.go:89] ID: 33 got omap values: (pool="replicapool", namespace="", name="csi.volume.1d186650-ea71-49b8-b98e-e68a22cb9f94"): map[csi.imageid:580777684f59f7 csi.imagename:csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94 csi.volname:pvc-f3173c10-61d4-4e43-b60f-a7b961e689fb csi.volume.owner:default]
W0721 05:57:46.709094       1 controllerserver.go:1904] ID: 33 metadata .rbd.csi.ceph.com/clientAddress/c1-m02 is not set for image replicapool/csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94
I0721 05:57:46.709262       1 utils.go:348] ID: 33 GRPC response: {}
```

```
[pm@dhcp53-115 rbd]$ k get po --field-selector spec.nodeName=c1-m02
NAME                       READY   STATUS    RESTARTS   AGE
rbd-demo-5d6f7d499-qc7p6   1/1     Running   0          27s
[pm@dhcp53-115 rbd]$
```

```
bash-5.1$ rbd status replicapool/csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94
Watchers:
	watcher=10.244.1.1:0/760543125 client.5771468 cookie=18446462598732840961
bash-5.1$ rbd image-meta ls replicapool/csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94
There are 5 metadata on this image:

Key                                     Value
.rbd.csi.ceph.com/clientAddress/c1-m02  10.244.1.1
.rbd.csi.ceph.com/clientAddress/c1-m03  192.168.39.16
csi.storage.k8s.io/pv/name              pvc-f3173c10-61d4-4e43-b60f-a7b961e689fb
csi.storage.k8s.io/pvc/name             rbd-pvc
csi.storage.k8s.io/pvc/namespace        default
```

<hr>

4. Node Recovery

```
[pm@dhcp53-115 rbd]$ k get node
NAME     STATUS   ROLES           AGE   VERSION
c1       Ready    control-plane   21d   v1.32.0
c1-m02   Ready    worker          21d   v1.32.0
c1-m03   Ready    worker          12d   v1.32.0

[pm@dhcp53-115 rbd]$ kubectl patch nodes c1-m03 -p '{"spec":{"taints":[]}}' --type=merge
node/c1-m03 patched
```

delete pod from node c1-m02 and pod is scheduled on recoverd node (c1-m03)

```
[pm@dhcp53-115 rbd]$ k delete po rbd-demo-5d6f7d499-qc7p6
pod "rbd-demo-5d6f7d499-qc7p6" deleted

[pm@dhcp53-115 rbd]$ k get po --field-selector spec.nodeName=c1-m02
No resources found in default namespace.

[pm@dhcp53-115 rbd]$ k get po --field-selector spec.nodeName=c1-m03
NAME                       READY   STATUS    RESTARTS   AGE
rbd-demo-5d6f7d499-c5qsk   1/1     Running   0          29s
[pm@dhcp53-115 rbd]$
```

/ControllerUnpublishVolume
```
I0721 06:06:32.659327       1 utils.go:341] ID: 35 GRPC call: /csi.v1.Controller/ControllerUnpublishVolume
I0721 06:06:32.660392       1 utils.go:342] ID: 35 GRPC request: {"node_id":"c1-m02","secrets":"***stripped***","volume_id":"0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94"}
I0721 06:06:32.687860       1 omap.go:89] ID: 35 got omap values: (pool="replicapool", namespace="", name="csi.volume.1d186650-ea71-49b8-b98e-e68a22cb9f94"): map[csi.imageid:580777684f59f7 csi.imagename:csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94 csi.volname:pvc-f3173c10-61d4-4e43-b60f-a7b961e689fb csi.volume.owner:default]
I0721 06:06:32.834924       1 controllerserver.go:1816] ID: 35 node is not marked as out-of-service, removed metadata .rbd.csi.ceph.com/clientAddress/c1-m02 from image replicapool/csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94 if it existed
I0721 06:06:32.835106       1 utils.go:348] ID: 35 GRPC response: {}
```

/ControllerPublishVolume
```
I0721 06:06:33.253588       1 utils.go:341] ID: 36 GRPC call: /csi.v1.Controller/ControllerPublishVolume
I0721 06:06:33.255092       1 utils.go:342] ID: 36 GRPC request: {"node_id":"c1-m03","secrets":"***stripped***","volume_capability":{"access_mode":{"mode":"SINGLE_NODE_WRITER"},"block":{}},"volume_context":{"clusterID":"rook-ceph","imageFeatures":"layering","imageFormat":"2","imageName":"csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94","journalPool":"replicapool","pool":"replicapool","storage.kubernetes.io/csiProvisionerIdentity":"1753075775496-228-rook-ceph.rbd.csi.ceph.com"},"volume_id":"0001-0009-rook-ceph-0000000000000005-1d186650-ea71-49b8-b98e-e68a22cb9f94"}
I0721 06:06:33.260773       1 omap.go:89] ID: 36 got omap values: (pool="replicapool", namespace="", name="csi.volume.1d186650-ea71-49b8-b98e-e68a22cb9f94"): map[csi.imageid:580777684f59f7 csi.imagename:csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94 csi.volname:pvc-f3173c10-61d4-4e43-b60f-a7b961e689fb csi.volume.owner:default]
I0721 06:06:35.036984       1 cephcmds.go:109] ID: 36 command succeeded: ceph [osd blocklist rm 192.168.39.16 --id=csi-rbd-provisioner --keyfile=***stripped*** -m=10.105.101.174:6789]
I0721 06:06:35.037375       1 cephcmds.go:358] ID: 36 unblocked IP "192.168.39.16" successfully
I0721 06:06:35.093087       1 utils.go:348] ID: 36 GRPC response: {}
```

```
bash-5.1$ rbd image-meta ls replicapool/csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94
There are 4 metadata on this image:

Key                                     Value
.rbd.csi.ceph.com/clientAddress/c1-m03  192.168.39.16
csi.storage.k8s.io/pv/name              pvc-f3173c10-61d4-4e43-b60f-a7b961e689fb
csi.storage.k8s.io/pvc/name             rbd-pvc
csi.storage.k8s.io/pvc/namespace        default

bash-5.1$ ceph osd blocklist ls
listed 0 entries

bash-5.1$ rbd status replicapool/csi-vol-1d186650-ea71-49b8-b98e-e68a22cb9f94
Watchers:
	watcher=192.168.39.16:0/1440394271 client.5772054 cookie=18446462598732840961
```

<hr>

5. ControllerPublish Secrets not in StorageClass & ceph-csi-config ConfigMap

```
I0721 06:47:31.582722       1 utils.go:341] ID: 24 GRPC call: /csi.v1.Controller/ControllerUnpublishVolume
I0721 06:47:31.582865       1 utils.go:342] ID: 24 GRPC request: {"node_id":"c1-m03","volume_id":"0001-0009-rook-ceph-0000000000000005-fc0f678c-428c-47af-873d-7af5a5223fab"}
W0721 06:47:31.583786       1 controllerserver.go:1785] ID: 24 controller publish secret not found: controller publish secret name or namespace is empty in csi config file for cluster rook-ceph
I0721 06:47:31.583880       1 utils.go:348] ID: 24 GRPC response: {}

I0721 06:50:07.569280       1 utils.go:341] ID: 25 GRPC call: /csi.v1.Controller/ControllerPublishVolume
I0721 06:50:07.569484       1 utils.go:342] ID: 25 GRPC request: {"node_id":"c1-m03","volume_capability":{"access_mode":{"mode":"SINGLE_NODE_WRITER"},"block":{}},"volume_context":{"clusterID":"rook-ceph","imageFeatures":"layering","imageFormat":"2","imageName":"csi-vol-fc0f678c-428c-47af-873d-7af5a5223fab","journalPool":"replicapool","pool":"replicapool","storage.kubernetes.io/csiProvisionerIdentity":"1753077229667-7755-rook-ceph.rbd.csi.ceph.com"},"volume_id":"0001-0009-rook-ceph-0000000000000005-fc0f678c-428c-47af-873d-7af5a5223fab"}
W0721 06:50:07.569884       1 controllerserver.go:1883] ID: 25 controller publish secret not found: controller publish secret name or namespace is empty in csi config file for cluster rook-ceph
I0721 06:50:07.570052       1 utils.go:348] ID: 25 GRPC response: {}
```

<hr>

6. ControllerPublish Secrets not in StorageClass but in ceph-csi-config ConfigMap

```
I0721 06:57:34.434549       1 utils.go:341] ID: 26 GRPC call: /csi.v1.Controller/ControllerUnpublishVolume
I0721 06:57:34.434679       1 utils.go:342] ID: 26 GRPC request: {"node_id":"c1-m03","volume_id":"0001-0009-rook-ceph-0000000000000005-fc0f678c-428c-47af-873d-7af5a5223fab"}
I0721 06:57:34.518576       1 omap.go:89] ID: 26 got omap values: (pool="replicapool", namespace="", name="csi.volume.fc0f678c-428c-47af-873d-7af5a5223fab"): map[csi.imageid:5814a64c577f1f csi.imagename:csi-vol-fc0f678c-428c-47af-873d-7af5a5223fab csi.volname:pvc-bf618a64-fc40-422f-9cf6-35f1e236218b csi.volume.owner:default]
I0721 06:57:34.634808       1 controllerserver.go:1816] ID: 26 node is not marked as out-of-service, removed metadata .rbd.csi.ceph.com/clientAddress/c1-m03 from image replicapool/csi-vol-fc0f678c-428c-47af-873d-7af5a5223fab if it existed
I0721 06:57:34.635026       1 utils.go:348] ID: 26 GRPC response: {}

I0721 06:58:02.498739       1 utils.go:341] ID: 27 GRPC call: /csi.v1.Controller/ControllerPublishVolume
I0721 06:58:02.499086       1 utils.go:342] ID: 27 GRPC request: {"node_id":"c1-m03","volume_capability":{"access_mode":{"mode":"SINGLE_NODE_WRITER"},"block":{}},"volume_context":{"clusterID":"rook-ceph","imageFeatures":"layering","imageFormat":"2","imageName":"csi-vol-fc0f678c-428c-47af-873d-7af5a5223fab","journalPool":"replicapool","pool":"replicapool","storage.kubernetes.io/csiProvisionerIdentity":"1753077229667-7755-rook-ceph.rbd.csi.ceph.com"},"volume_id":"0001-0009-rook-ceph-0000000000000005-fc0f678c-428c-47af-873d-7af5a5223fab"}
I0721 06:58:02.509900       1 omap.go:89] ID: 27 got omap values: (pool="replicapool", namespace="", name="csi.volume.fc0f678c-428c-47af-873d-7af5a5223fab"): map[csi.imageid:5814a64c577f1f csi.imagename:csi-vol-fc0f678c-428c-47af-873d-7af5a5223fab csi.volname:pvc-bf618a64-fc40-422f-9cf6-35f1e236218b csi.volume.owner:default]
W0721 06:58:02.619188       1 controllerserver.go:1904] ID: 27 metadata .rbd.csi.ceph.com/clientAddress/c1-m03 is not set for image replicapool/csi-vol-fc0f678c-428c-47af-873d-7af5a5223fab
I0721 06:58:02.620353       1 utils.go:348] ID: 27 GRPC response: {}
```

iPraveenParihar committed:
helm: add attacher sidecar for CephFS deployment

This commit updates helm templates for CephFS deployment to include
csi-attacher sidecar.

For the Fencing node feature (--enable-fencing) CephFS implements
ControllerPublishVolume and ControllerUnpublishVolume RPCs which
requires attacher sidecar to call the RPCs.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
deploy: add attacher sidecar for CephFS deployment

This commit updates deploy/ templates for CephFS deployment to include
csi-attacher sidecar.

For the Fencing node feature (--enable-fencing) CephFS implements
ControllerPublishVolume and ControllerUnpublishVolume RPCs which
requires attacher sidecar to call the RPCs.

Signed-off-by: Praveen M <m.praveen@ibm.com>

nixpanic commented on internal/util/util.go:
```diff
@@ -283,3 +285,29 @@ func GetVolumeContext(parameters map[string]string) map[string]string {
 	return volumeContext
 }
+
+func ParseClientIP(addr string) (string, error) {
```
like other functions that are moved to more common usage, comments should be added

nixpanic commented on cmd/cephcsi.go:
```diff
@@ -177,13 +171,13 @@ func getDriverName() string {
 	// select driver name based on volume type
 	switch conf.Vtype {
-	case rbdType:
+	case util.RBDType:
```
The commit message has a type. Why is this needed/useful? **THAT** should be included in the commit message.

nixpanic commented on internal/util/util.go:
```diff
@@ -36,6 +36,14 @@ import (
 	mount "k8s.io/mount-utils"
 )
 
+const (
```
types for what? please add comments

nixpanic commented on internal/csi-common/driver.go:
```diff
@@ -35,6 +35,8 @@ type CSIDriver struct {
 	// ceph clusters across CSI instances, to differentiate omap names per CSI instance.
 	instance string
 
+	enableFencing bool // set client address metadata on the volume
```
like other comments, place it above the attribute

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -1679,11 +1679,19 @@ func (cs *ControllerServer) ControllerExpandVolume(
+	// If the CSI driver is not running on Kubernetes, we do not need to
+	// unfence the node, so we return immediately.
+	// This is because the unfencing logic is specific to Kubernetes environments.
+	if !k8s.RunsOnKubernetes() {
```
not only on Kubernetes, also if fencing isn't enabled this function should return early.

iPraveenParihar replied:
early return is done in `fenceNode/unfenceNode` methods.

Today, we just have fencing operations but we might have other things in future
For example, removing nodeId:userId mapping metadata in ControllerUnpublishVolume.

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.31

Madhu-1 commented on charts/ceph-csi-cephfs/templates/provisioner-clusterrole.yaml:
```diff
@@ -65,6 +65,14 @@ rules:
     resources: ["volumesnapshotcontents"]
     verbs: ["create", "get", "list", "watch", "update", "delete", "patch"]
 {{- end -}}
+{{- if .Values.provisioner.attacher.enabled }}
```
we might not need this one right?

iPraveenParihar replied:
attacher is only required if `--enable-fencing` is set true and not always?

Madhu-1 replied:
and for other features also. If we dont deploy the attacher and set the attachRequired in the csidriver CR will it work? am fine if it works

iPraveenParihar replied:
yea, missed configuring that.

We can either configure the CSIDriver template Or deploy attacher always?
```
attachRequired: {{ .Values.CSIDriver.attachRequired }}  # true by default
```

@Madhu-1 ^

Madhu-1 replied:
Yes sounds good 👍🏻

Madhu-1 replied:
Lets make this as configurable

iPraveenParihar replied:
Done

Madhu-1 commented on charts/ceph-csi-cephfs/templates/provisioner-deployment.yaml:
```diff
@@ -203,6 +203,41 @@ spec:
               mountPath: /csi
           resources:
 {{ toYaml .Values.provisioner.snapshotter.resources | indent 12 }}
+{{- if .Values.provisioner.attacher.enabled }}
```
we want attacher to be deployed always

iPraveenParihar replied:
attacher is only required if --enable-fencing is set true and not always?
Anyhow, the default value for attacher.enabled is `true`.

Madhu-1 commented on charts/ceph-csi-cephfs/values.yaml:
```diff
@@ -174,6 +174,8 @@ provisioner:
       # unavailable during the update process.
       maxUnavailable: 50%
+  # if fstype is not specified in storageclass, ext4 is default
+  defaultFSType: ext4
```
why this is required for cephfs?

iPraveenParihar replied:
Oops, didn't think through
not required for CephFs, removing it.

Madhu-1 reviewed (commented):
small nits, LGTM. we can do the enhancements as we test the feature to larger extend

Madhu-1 commented on charts/ceph-csi-cephfs/values.yaml:
```diff
@@ -77,6 +77,7 @@ logSlowOperationInterval: 30s
 # Changes will only occur if the fsType is defined and the persistent volume's
 # accessModes contains ReadWriteOnce.
 CSIDriver:
+  attachRequired: true
```
lets not add extra flag use attacher.enabled flag itself for it.

iPraveenParihar replied:
done

Rakshith-R reviewed (approved):
Thanks
@mergifyio queue

Rakshith-R commented:
@mergifyio queue

iPraveenParihar commented:
/retest ci/centos/k8s-e2e-external-storage/1.32

remram44 commented:
What's the upgrade path for this? CSIDriver is immutable, and it is not even mentioned in [your upgrade docs](https://github.com/ceph/ceph-csi/blob/devel/docs/ceph-csi-upgrade.md#upgrading-from-v314-to-v315).

pschichtel commented:
@remram44 simplest path would be to disable the attacher, the alternative is to delete your CSIDriver object and recreate it with the attacher enabled. I don't think there are any other options.

remram44 commented:
Delete before deploying the new manifests is easy, does that work? Is it safe?

Should it perhaps be listed in your upgrade steps?

iPraveenParihar commented:
@remram44, We did not mention this feature in the 3.15 release notes because we encountered a bug at later stage and the fix could not be included in that release. The issue will be resolved in the 3.15.1 release, and the proper upgrade steps will be documented there.

remram44 commented:
I see! You could mention *that* in the release notes though, so people like me don't update to 3.15.0 (too late for me though)
